KEP-9270: Introduce configuration for the MultiKueue Incremental Dispatcher (#10877)
kubernetes-sigs/kueue: keps/9270-multikueue-incremental-step-size/README.md

Reviewer: Nit: this line still mis-renders.
(Perhaps you have some auto-formatter doing it? If so, consider adding blank lines to prevent it).

 
# KEP-9270: MultiKueue Incremental Dispatcher Step Size
<!-- toc -->
- [Summary](#summary)
- [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
- [Proposal](#proposal)
  - [Risks and Mitigations](#risks-and-mitigations)
- [Design Details](#design-details)
  - [API Changes](#api-changes)
  - [Execution Loop](#execution-loop)
  - [Test Plan](#test-plan)
    - [Prerequisite testing updates](#prerequisite-testing-updates)
    - [Unit tests](#unit-tests)
    - [Integration tests](#integration-tests)
    - [e2e tests](#e2e-tests)
  - [Graduation Criteria](#graduation-criteria)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
<!-- /toc -->

## Summary

This proposal introduces a new configuration field called `stepSize` to the MultiKueue Incremental Dispatcher. This setting will control the number of worker clusters that the Manager Cluster will attempt to dispatch a workload to concurrently.

## Motivation

> Currently, when MultiKueue attempts to schedule a workload across multiple worker clusters, the Incremental Dispatcher has a limitation. If it checks clusters one by one, the process is too slow and delays job execution. If it attempts to dispatch to all clusters simultaneously, it generates unnecessary API traffic, increases the load on the Manager Cluster, and can lead to race conditions where multiple clusters try to admit the same job at the same time.

Reviewer: Nit: I feel this paragraph is somewhat confusing.

I think you captured the _tradeoff_ well (between slow dispatching & high resource usage).
However, that tradeoff _does not yet exist_, as we currently hard-code `batchSize := 3`.
I think this paragraph sounds as if this was already variable.

Also, when you write "the Incremental Dispatcher has a limitation", it remains unclear what's the "limitation" specifically.

Introducing a `stepSize` allows cluster administrators to configure a "batch size" for this process, finding the perfect balance between speed and system stability.

### Goals

* Add a `stepSize` parameter to the `MultiKueueConfig` API.
* Update the Workload Controller's dispatching loop to process worker clusters in batches defined by the `stepSize`.

> * Ensure that if a workload is admitted in a batch, the remaining pending proxy workloads in that batch are cleanly removed.

Reviewer: 1. Why only in that batch?
2. Isn't this already done, [here](https://github.com/kubernetes-sigs/kueue/blob/3a3d405ff0596aa4d3f433c90a73b570b59b2e94/pkg/controller/admissionchecks/multikueue/workload.go#L460-L469)?

(I'm not sure if it's worth mentioning as a goal).

### Non-Goals

* We will not change the way worker clusters are scored or prioritized.
* We will not alter the `AllAtOnce` dispatcher behavior, as this proposal strictly targets the Incremental Dispatcher.

## Proposal

We propose updating the `MultiKueueConfig` API to accept an optional integer for the step size, and updating the Workload Controller to respect this batch limit when iterating through the list of available worker clusters.

### Risks and Mitigations

**Risk:** An administrator might set the `stepSize` to an extremely high number (e.g., equal to the total number of clusters), which effectively recreates the heavy load issues of the `AllAtOnce` dispatcher and increases the chance of race conditions where multiple clusters admit the job at the exact same time.

> **Mitigation:** Provide clear API documentation regarding best practices. The Workload Controller's conflict resolution logic (which deletes duplicate proxy workloads once one is admitted) will naturally handle race conditions, just as it already does for the `AllAtOnce` strategy. 

Reviewer: Nit: please separate them by an empty line, for better rendering.

> **Mitigation:** Provide clear API documentation regarding best practices. The Workload Controller's conflict resolution logic (which deletes duplicate proxy workloads once one is admitted) will naturally handle race conditions, just as it already does for the `AllAtOnce` strategy. 

Reviewer: > (which deletes duplicate proxy workloads once one is admitted)

Nit: I'd remove this part because this is not the only race condition threatening us.
See e.g. [KEP-8303](https://github.com/kubernetes-sigs/kueue/tree/main/keps/8303-multikueue-orchestrated-preemption) which deals with another kind of race condition.

(Still, the rest of the sentence makes sense: whatever prevention mechanisms we have, they will still act).

> **Mitigation:** Provide clear API documentation regarding best practices. The Workload Controller's conflict resolution logic (which deletes duplicate proxy workloads once one is admitted) will naturally handle race conditions, just as it already does for the `AllAtOnce` strategy. 

Reviewer: > regarding best practices

Out of curiosity, what "best practices" do you have in mind?
Is there any number (constant, or dependent on N = number of clusters) which you wouldn't recommend exceeding?

Asking because I don't know. Maybe N/2 is still reasonable for some use cases? Maybe even more? (Presumably, some users are happy even with `AllAtOnce`).

Author: > Out of curiosity, what "best practices" do you have in mind?

Since you are providing them with a new configuration, they need to know how to use it safely. The 'best practices' that you will document will typically include the following:

**Avoiding extreme values:**  Warning them against setting the `stepSize` to a massive number (such as the total number of clusters they have). You will clarify that doing so defeats the purpose of this feature and recreates the exact issues of high API load and race conditions

## Design Details

### API Changes

> We will update the `MultiKueueConfigSpec` struct to include the new field. This change will be backwards compatible by defaulting the step size to `1` (which matches the current one-by-one behavior).

Reviewer: > defaulting the step size to `1` (which matches the current one-by-one behavior).

Hmm, is the current behavior one-by-one?

https://github.com/kubernetes-sigs/kueue/blob/3a3d405ff0596aa4d3f433c90a73b570b59b2e94/pkg/controller/workloaddispatcher/incrementaldispatcher.go#L180

Author: You're right. I knew it was currently 3, but I incorrectly assumed we should just default to the minimum value. Sticking with 3 makes much more sense so we maintain the exact same behavior for users when they upgrade.

```go
type MultiKueueConfigSpec struct {
    // Other existing fields...

    // StepSize defines the number of worker clusters the Incremental 
    // Dispatcher will query at the same time. 
    // Minimum value is 1. If not set, it defaults to 1.
    // +optional
    // +kubebuilder:default=1
    // +kubebuilder:validation:Minimum=1

>     StepSize *int32 `json:"stepSize,omitempty"`

Reviewer: With this proposal, if a user wants to use incremental dispatcher with step 4, they'll have to say:

```yaml
# in the global Configuration object for Kueue
multiKueue:
  dispatcherName: kueue.x-k8s.io/multikueue-dispatcher-incremental

# in the MultiKueueConfig (referenced by the MK AC, referenced by the manager CQ)
spec:
  stepSize: 4
```

This _might_ be best - but I'd like to discuss:

1. Do we really want step size configured in a very distant place from the choice of the dispatcher?
   (This would be justified if we needed to support different step sizes for different CQs.
   However, note that the choice "Incremental or AllAtOnce or Custom" is already global.
   So maybe we have no reason to stand out?)

2. Regardless of the answer to (1), I think I'd prefer the field name to be more self-explaining.
   E.g. `IncrementalDispatcherStepSize`.

BTW, if we choose to keep both fields next to each other, I'd also add validation that "step size" can be set _only if_ "incremental" is actually chosen. (This can be done with kubebuilder rules).

Author: I agree with these points. Sticking the step size next to the dispatcher globally makes much more sense.

I'll update the KEP to:

1-Move the field to the global Configuration struct.

2-Rename it to `IncrementalDispatcherStepSize`.

3-Add kubebuilder validation so it can only be used with the incremental dispatcher.

}
```

### Execution Loop

> The core logic change will take place inside the MultiKueue Workload Controller's reconciliation loop:

Reviewer: Hmm, it seems that we need less changes than that.
IIUC the batching logic already exists in [this file](https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/workloaddispatcher/incrementaldispatcher.go); we just need to replace the hard-coded `batchSize := 3` with your proposed custom parameter.

Unless I'm confusing things.

1. When a workload is ready for dispatch, the controller reads the `stepSize` from the `MultiKueueConfig`.
2. The controller takes the full list of available worker clusters and divides them into groups (batches) equal to the `stepSize`.
3. The controller loops through the first batch, creating proxy workloads on those specific worker clusters simultaneously.
4. If a cluster in the batch admits the job, the controller immediately deletes the proxy workloads from the other clusters in that same batch and stops the loop.
5. If no cluster in the batch admits the job, the controller moves on to the next batch of clusters.

### Test Plan

 I/we understand the owners of the involved components may require updates to existing tests to make this code solid enough prior to committing the changes necessary to implement this enhancement.

#### Prerequisite testing updates

No major prerequisite testing updates are required. Current mock worker cluster frameworks in EnvTest are sufficient.

#### Unit tests

- `pkg/controller/core/workload_controller_test.go`: `<date>` - Verify that the worker cluster list is correctly divided into chunks matching the `stepSize`.

> - `pkg/controller/core/workload_controller_test.go`: `<date>` - Verify that the default value remains `1` if the user does not specify a `stepSize`.

Reviewer: ```suggestion
- `pkg/controller/core/workload_controller_test.go`: Verify that the worker cluster list is correctly divided into chunks matching the `stepSize`.
- `pkg/controller/core/workload_controller_test.go`: Verify that the default value remains `1` if the user does not specify a `stepSize`.
```

#### Integration tests

- `TestIncrementalDispatcherStepSize`: Simulate a Manager Cluster with 5 Worker Clusters. Set `stepSize` to `2`. Submit a workload and verify via the API server that exactly 2 proxy workloads are created in the first step. Simulate a rejection on the first 2 clusters, and verify that the controller correctly creates the next 2 proxy workloads in the second step.

#### e2e tests

Standard E2E multi-cluster tests will be updated to include an iteration where `MultiKueueConfig` is deployed with a `stepSize > 1` to ensure workloads execute successfully end-to-end without duplication.

### Graduation Criteria

> * **Alpha:** * API fields added.

Reviewer: ```suggestion
* **Alpha:**
    * API fields added.
```

and same for "Beta" below.

Reviewer: Thanks for fixing here - but it's still there for "Beta".

    * Workload Controller updated to support batching.
    * Unit tests and EnvTest integration tests passing.
* **Beta:** * Gather feedback from the community.
    * E2E tests added and passing consistently.
* **Stable:**
    * Feature has been in Beta for at least one release cycle with no major bugs reported.

## Implementation History

- 2026-05-01: KEP proposed and initial draft created.

## Drawbacks

This adds slight code complexity to the existing Incremental Dispatcher loop and introduces one additional configuration parameter for administrators to manage and tune.

> ```

Reviewer: ```suggestion
```

Introducing a `stepSize` allows cluster administrators to configure a "batch size" for this process, finding the perfect balance between speed and system stability.

### Goals

* Add a `stepSize` parameter to the `MultiKueueConfig` API.
* Update the Workload Controller's dispatching loop to process worker clusters in batches defined by the `stepSize`.

### Non-Goals

* We will not change the way worker clusters are scored or prioritized.
* We will not alter the `AllAtOnce` dispatcher behavior, as this proposal strictly targets the Incremental Dispatcher.

## Proposal

We propose updating the `MultiKueueConfig` API to accept an optional integer for the step size, and updating the Workload Controller to respect this batch limit when iterating through the list of available worker clusters.

### Risks and Mitigations

> **Risk:**An administrator might set the `stepSize` to an extremely high number (e.g., equal to the total number of clusters), which effectively recreates the heavy load issues of the `AllAtOnce` dispatcher and increases the chance of race conditions where multiple clusters admit the job at the exact same time.

Reviewer: `**Risk**` misrenders, probably due to a missing space after that.

### Risks and Mitigations

**Risk:** An administrator might set the `stepSize` to an extremely high number (e.g., equal to the total number of clusters), which effectively recreates the heavy load issues of the `AllAtOnce` dispatcher and increases the chance of race conditions where multiple clusters admit the job at the exact same time.

**Mitigation:** Provide clear API documentation regarding best practices. The Workload Controller's conflict resolution logic will naturally handle race conditions, just as it already does for the `AllAtOnce` strategy.

## Design Details

### API Changes

We will update the global `Configuration` API (specifically the `MultiKueue` struct) to include the new field, rather than placing it in `MultiKueueConfigSpec`. This keeps the step size configuration co-located with the `DispatcherName` selection.

The field will be named `IncrementalDispatcherStepSize` for clarity. To ensure backwards compatibility and preserve existing system behavior, the default value will be `3`.

> We will also use Kubebuilder validation markers (e.g., `CEL` rules or `+union` discriminators depending on the struct setup) to enforce that this field can only be populated if `DispatcherName` is set to the incremental dispatcher.

Reviewer: Nit: I think the "struct setup" is already known.

}
```

### Execution Loop
The core logic change is minimal because the batching mechanism is already implemented in the Incremental Dispatcher (`pkg/controller/workloaddispatcher/incrementaldispatcher.go`).

The update will simply modify the existing logic to:
1. Read the `stepSize` value from the `MultiKueueConfig`.
2. Replace the hardcoded `batchSize := 3` variable with this dynamically configured `stepSize` parameter.

The existing loop and conflict resolution mechanisms will naturally handle the rest of the execution based on this new size, meaning no major architectural changes are required to the dispatching loop itself.

### Test Plan

 I/we understand the owners of the involved components may require updates to existing tests to make this code solid enough prior to committing the changes necessary to implement this enhancement.

#### Prerequisite testing updates

No major prerequisite testing updates are required. Current mock worker cluster frameworks in EnvTest are sufficient.

#### Unit tests

- `pkg/controller/core/workload_controller_test.go`: Verify that the worker cluster list is correctly divided into chunks matching the `stepSize`.

> - `pkg/controller/core/workload_controller_test.go`: Verify that the default value remains `1` if the user does not specify a `stepSize`.

Reviewer: ```suggestion
- `pkg/controller/core/workload_controller_test.go`: Verify that the default value remains `3` if the user does not specify a `stepSize`.
```

#### Integration tests

- `TestIncrementalDispatcherStepSize`: Simulate a Manager Cluster with 5 Worker Clusters. Set `stepSize` to `2`. Submit a workload and verify via the API server that exactly 2 proxy workloads are created in the first step. Simulate a rejection on the first 2 clusters, and verify that the controller correctly creates the next 2 proxy workloads in the second step.

#### e2e tests

Standard E2E multi-cluster tests will be updated to include an iteration where `MultiKueueConfig` is deployed with a `stepSize > 1` to ensure workloads execute successfully end-to-end without duplication.

### Graduation Criteria

* **Alpha:**
    * API fields added.
    * Workload Controller updated to support batching.
    * Unit tests and EnvTest integration tests passing.

> * **Beta:** * Gather feedback from the community.

Reviewer: ```suggestion
* **Beta:**
    * Gather feedback from the community.
```

Reviewer: This still misrenders; perhaps you have some auto-formatter doing it?
(If so, consider adding blank lines to prevent this).

**Mitigation:** Provide clear API documentation regarding best practices. The Workload Controller's conflict resolution logic will naturally handle race conditions, just as it already does for the `AllAtOnce` strategy.

## Design Details

### API Changes

We will update the global `Configuration` API (specifically the `MultiKueue` struct) to include a new configuration struct, rather than placing it in `MultiKueueConfigSpec`. This keeps the incremental dispatcher configuration co-located with the `DispatcherName` selection.

The field will be named `IncrementalDispatcherConfig` and will contain a `StepSize` field. To ensure backwards compatibility and preserve existing system behavior, the default value for `StepSize` will be `3`.

```go
type MultiKueue struct {
       // Other existing fields...

      // DispatcherName defines the type of dispatcher to use.
      DispatcherName *string `json:"dispatcherName,omitempty"`

> 

Reviewer 2: ```suggestion
```

Can you exclude already existing fields from API changes?

      // IncrementalDispatcherConfig contains the configuration for the incremental dispatcher.
      // This field is only valid when DispatcherName is set to the incremental dispatcher.
      // Note: This field is ignored when the MultiKueueIncrementalDispatcher feature gate is disabled.

>       // +optional

Reviewer 2: This is just a side note. This API is a configuration. So, these markers don't work for anything. Then we have to implement validations and defaulting manually.

      IncrementalDispatcherConfig *IncrementalDispatcherConfig `json:"incrementalDispatcherConfig,omitempty"`
}

type IncrementalDispatcherConfig struct {
      // StepSize defines the number of worker clusters the Incremental 
      // Dispatcher will query simultaneously. 

>       // Minimum value is 1. If not set, it defaults to 3.

Reviewer 2: Also, this is just a side note. This defaulting should be implemented by manual defaulter implementations: https://github.com/kubernetes-sigs/kueue/blob/5446b2a38eeec3615d95b3e06ecd0ec311ce7d38/apis/config/v1beta2/defaults.go#L68

- [Implementation History](#implementation-history)
- [Possible Follow-ups](#possible-follow-ups)
- [Drawbacks](#drawbacks)
<!-- /toc -->

## Summary

This proposal introduces a new configuration structure for the MultiKueue Incremental Dispatcher, starting with a stepSize field. This setting will control the number of worker clusters that the Manager Cluster will attempt to dispatch a workload to concurrently.

## Motivation

Currently, the MultiKueue Incremental Dispatcher has a hardcoded batch size `batchSize := 3` when attempting to schedule a workload across multiple worker clusters. The primary limitation is that this value is not configurable. For administrators managing a massive fleet of worker clusters, a fixed batch size of 3 is too slow and delays job execution. Conversely, if administrators were to theoretically increase this to query all clusters simultaneously to speed things up, it would generate heavy API traffic and increase the risk of race conditions. By making this step size configurable, administrators can find the optimal balance between dispatch speed and Manager Cluster resource usage for their specific environment.

> Introducing a `stepSize` allows cluster administrators to configure a "batch size" for this process, finding the perfect balance between speed and system stability.

Reviewer 3: ### 
```suggestion
Currently, the MultiKueue Incremental Dispatcher has a hardcoded batch size `batchSize := 3` when attempting to schedule a workload across multiple worker clusters. 

Sometimes administrators need to configure the parameter in order to:
1. avoid unnecessary job execution delays when there are many clusters (by increasing the stepSize)
2. use more granular prioritization of the clusters (by setting stepSize=1)
3. tune the api-server traffic and load on the manager cluster
```
nit, in order to:
1. make it less verbose and to-the-point
2. avoid the overselling statements like " finding the perfect balance"
3. add the use case for more granular prioritization of the clusters
4. rephrase the api-server traffiic point slightly

## Design Details

### API Changes

We will update the global `Configuration` API (specifically the `MultiKueue` struct) to include a new configuration struct, rather than placing it in `MultiKueueConfigSpec`. This keeps the incremental dispatcher configuration co-located with the `DispatcherName` selection.

The field will be named `IncrementalDispatcherConfig` and will contain a `StepSize` field. To ensure backwards compatibility and preserve existing system behavior, the default value for `StepSize` will be `3`.

We will also use Kubebuilder validation markers (e.g., CEL rules) to enforce that this struct can only be populated if `DispatcherName` is set to the incremental dispatcher.

```go
type MultiKueue struct {
       // Other existing fields...

      // DispatcherName defines the type of dispatcher to use.
      DispatcherName *string `json:"dispatcherName,omitempty"`

      // IncrementalDispatcherConfig contains the configuration for the incremental dispatcher.
      // This field is only valid when DispatcherName is set to the incremental dispatcher.

>       // Note: This field is going to be ignored when the MultiKueueIncrementalDispatcher feature gate is disabled.

Reviewer 3: ```suggestion
      // Note: This field is ignored when the MultiKueueIncrementalDispatcher feature gate is disabled.
```
nit, assume the reader is a future user or developer, so "going to" will not be relevant soon.

      // +optional
      IncrementalDispatcherConfig *IncrementalDispatcherConfig `json:"incrementalDispatcherConfig,omitempty"`
}

type IncrementalDispatcherConfig struct {
      // StepSize defines the number of worker clusters the Incremental 
      // Dispatcher will query simultaneously. 
      // Minimum value is 1. If not set, it defaults to 3.
      // +optional
      // +kubebuilder:default=3

>       // +kubebuilder:validation:Minimum=1

Reviewer 3: nit, but the kubebuilder markers don't matter for ConfigAPI , because the kubebuilder markers are only used for schema generation, but we don't generate the schema for the configmap api.,

      StepSize *int32 `json:"stepSize,omitempty"`
}
```

### Execution Loop
The core logic change is minimal because the batching mechanism is already implemented in the Incremental Dispatcher (`pkg/controller/workloaddispatcher/incrementaldispatcher.go`).

The update will simply modify the existing logic to:
1. Read the `stepSize` value from the `MultiKueueConfig`.
2. Replace the hardcoded `batchSize := 3` variable with this dynamically configured `stepSize` parameter.

> The existing loop and conflict resolution mechanisms will naturally handle the rest of the execution based on this new size, meaning no major architectural changes are required to the dispatching loop itself.

Reviewer 3: ```suggestion
### Implementation overview

The core logic change is minimal because the batching mechanism is already implemented in the Incremental Dispatcher, assuming the hard-coded value of stepSize=3.  
No architectural changes are required to the dispatching loop itself.
```
nit, but I think this section is too verbose, proposing to make it more concise.

#### Prerequisite testing updates

No major prerequisite testing updates are required. Current mock worker cluster frameworks in EnvTest are sufficient.

#### Unit tests

- `pkg/controller/core/workload_controller_test.go`: Verify that the worker cluster list is correctly divided into chunks matching the `stepSize`.
- `pkg/controller/core/workload_controller_test.go`: Verify that the default value remains `3` if the user does not specify a `stepSize`.

#### Integration tests

- `TestIncrementalDispatcherStepSize`: Simulate a Manager Cluster with 5 Worker Clusters. Set `stepSize` to `2`. Submit a workload and verify via the API server that exactly 2 proxy workloads are created in the first step. Simulate a rejection on the first 2 clusters, and verify that the controller correctly creates the next 2 proxy workloads in the second step.

#### e2e tests

Standard E2E multi-cluster tests will be updated to include an iteration where `MultiKueueConfig` is deployed with a `stepSize > 1` to ensure workloads execute successfully end-to-end without duplication.

### Graduation Criteria

* **Alpha:**

>     * API fields added.

Reviewer 3: nit, add that "MultiKueueIncrementalDispatcherConfig" is alpha (disabled by default)

for beta: "MultiKueueIncrementalDispatcherConfig" is beta (enabled by default)

for Stabke: "MultiKueueIncrementalDispatcherConfig" is locked (will be removed in N+2)

### API Changes

We will update the global `Configuration` API (specifically the `MultiKueue` struct) to include a new configuration struct, rather than placing it in `MultiKueueConfigSpec`. This keeps the incremental dispatcher configuration co-located with the `DispatcherName` selection.

The field will be named `IncrementalDispatcherConfig` and will contain a `StepSize` field. To ensure backwards compatibility and preserve existing system behavior, the default value for `StepSize` will be `3`.

We will also use Kubebuilder validation markers (e.g., CEL rules) to enforce that this struct can only be populated if `DispatcherName` is set to the incremental dispatcher.

```go
type MultiKueue struct {
// Other existing fields...
// DispatcherName defines the type of dispatcher to use.
DispatcherName *string `json:"dispatcherName,omitempty"`

// IncrementalDispatcherConfig contains the configuration for the incremental dispatcher.
// This field is only valid when DispatcherName is set to the incremental dispatcher.

> // Note: This field is going to be ignored when the MultiKueueStepSize feature gate is disabled.

Reviewer: Don't forget to update the FG name also here

// +optional

> IncrementalDispatcherConfig *IncrementalDispatcherConfig `json:"incrementalDispatcherConfig,omitempty"`

Reviewer: Nit: restore the indent
